day_5_supply_stacks/stack_parser.rb

```diff
@@ -0,0 +1,78 @@
+require 'minitest/autorun'
+
+class StackParser
+  def initialize(stack_input)
+    @stack_input = stack_input
+    @stack_rows = stack_rows
+  end
+
+  def stack_columns
+    raw_stack_columns = transpose_stack_rows
+    remove_empty_crates_from(raw_stack_columns)
+  end
+
+  private
+
+  attr_reader :stack_input, :stack_row
+
+  def transpose_stack_rows
+    rows = stack_rows.size
+    columns = stack_rows.max_by(&:size).size
+
+    raw_stack_columns = Array.new(columns) { Array.new(rows) }
+
+    stack_rows.each_with_index do |row, i|
+      row.each_with_index do |crate, j|
+        raw_stack_columns[j][i] = crate
+      end
+    end
+
+    raw_stack_columns
+  end
+
+  def remove_empty_crates_from(raw_stack_columns)
+    raw_stack_columns.map do |column|
+      column.delete_if { |crate| crate.nil? || crate.empty? }
+    end
+  end
+
+  def stack_rows
+    stack_input.map do |row|
+      row.each_char.each_slice(4).map do |crate|
+        crate.join('')
+             .gsub("[", "")
+             .gsub("]","")
+             .strip
```

Comment:
Another way to do this:

```ruby
crate.join.tr("[]", "").strip
```

`Array#join` joins the elements with an empty string by default, so you don't need to specify that. 

btw TIL how it really works: "With no argument, joins using the output field separator, `$,`" 

https://www.rubydoc.info/stdlib/core/Array:join

The variables starting with `$` in Ruby are called global variables. They are normally loathed upon (global state is bad, not just in Ruby, but in computer science in general - there is no scoping, anyone can change it anywhere, and you are fucked, with a hard to debug problem :)) On the upside, they are accessible from anywhere, so you don't need `attr_reader`s and such.

But anyways, by default, the value of the `$,` variable is `nil`, so invoking `Array#join` without parameters is the same as `join(nil)`, which has the same effect as `join('')`.

Looks like this is going to be deprecated, so I wouldn't rely on it - it's just an interesting tidbit :)

Reply:
As for `String#tr`, (`tr` is a shortcut of `translate` - not sure why did they shorten it, it's not typical to do it in Ruby).

Anyways, `String#tr` is a good idea to have in your toolbox, it's super powerful. Check it out:

https://ruby-doc.org/3.2.0/String.html#method-i-tr

Reply:
Awesome! Thanks. A new tool in my toolbox.